### PR TITLE
Update "Concepts"

### PR DESCRIPTION
**Pull request by adnanmorsy97**
I have fixed some spelling and grammar mistakes.

### CODE REVIEW

**Comment by adnanmorsy97** [comment]
@JohannesLichtenberger could you please merge it :) ?

File: concepts.md [hunk 1]
```diff
@@ -7,118 +7,118 @@ title: SirixDB - Architecture and Concepts
...
 
-The page-structure for one revision is depicted in the following figure:
+The page structure for one revision is depicted in the following figure:
 
 <div class="img_container">
 ![pageStructure](images/pageStructureOneRev.png){: style="max-width: 100%; height: auto; margin: 0em"}
 </div>
 
-**Each node and revision in SirixDB is referenced by a unique, stable identifier.** First, SirixDB has to find the revision by its revision number traversing a tree of indirect-pages. Addressing nodes is done in the same manner.
+**Each node and revision in SirixDB is referenced by a unique, stable identifier.** First, SirixDB has to find the revision by its revision number traversing a tree of indirect pages. Addressing nodes is done in the same manner.
```
Comment (mosheduminer) [inline concepts.md:50]: I think this should best be changed to `indirection pages`. What do you think @JohannesLichtenberger?

File: concepts.md [hunk 2]
```diff
@@ -7,118 +7,118 @@ title: SirixDB - Architecture and Concepts
...
-To support fast access to a `RevisionRootPage`, SirixDB stores a second file with just the offsets to specific revisions in a revisions-file,  which is read into main-memory on startup.
+: The `RevisionRootPage` is the main entry point to a revision. It stores the author-ID, an optional commit message, and a timestamp in the form of the unix epoch (milliseconds since 1970). Furthermore, it stores a reference to a `PathPage`, a `CASPage` (if it exists), a `NamePage`, and an `IndirectPage`. The indirect page is the entry point to the data stored in the leaf `RecordPages`. The right subtree of the `RevisionRootPage` started by the `IndirectPage` is the main entry point to our data stored in the leaf nodes, the RecordPages once again.
```
Comment (mosheduminer) [inline concepts.md:61]: I feel like this paragraph is unclear and could be improved, but I don't understand what is going on here. @JohannesLichtenberger?

File: concepts.md [hunk 3]
```diff
@@ -7,118 +7,118 @@ title: SirixDB - Architecture and Concepts
...
-To support fast access to a `RevisionRootPage`, SirixDB stores a second file with just the offsets to specific revisions in a revisions-file,  which is read into main-memory on startup.
+: The `RevisionRootPage` is the main entry point to a revision. It stores the author-ID, an optional commit message, and a timestamp in the form of the unix epoch (milliseconds since 1970). Furthermore, it stores a reference to a `PathPage`, a `CASPage` (if it exists), a `NamePage`, and an `IndirectPage`. The indirect page is the entry point to the data stored in the leaf `RecordPages`. The right subtree of the `RevisionRootPage` started by the `IndirectPage` is the main entry point to our data stored in the leaf nodes, the RecordPages once again. 
+To support fast access to a `RevisionRootPage`, SirixDB stores a second file with just the offsets to specific revisions in a revisions file, which is read into main-memory on startup.
```
Comment (mosheduminer) [inline concepts.md:62]: I think it should be left as `revisions-file`, consistent with the usage of `data-file` in this document, for example.

File: concepts.md [hunk 4]
```diff
@@ -7,118 +7,118 @@ title: SirixDB - Architecture and Concepts
...
+: The `PathPage` has references to `IndirectPages`, whereas each of the indirect pages is the root entry point to a user-defined path index. A unique index ID is also the reference offset in the path page to retrieve the respective path index/`IndirectPage` subtree root. SirixDB adds references to indirect pages, once a user creates path indexes. SirixDB stores an AVL-tree for each index in the leaf pages (the `RecordPage`s), which are referenced by the indirect pages. The index contains path nodes as keys as well as an array of record-identifiers in the values. Furthermore, a path page contains a reference to a `PathSummary` page.
```
Comment (mosheduminer) [inline concepts.md:65]: I think the term is `IndirectPage`, so perhaps the old way is preferable.

File: concepts.md [hunk 5]
```diff
@@ -7,118 +7,118 @@ title: SirixDB - Architecture and Concepts
...
 
 RecordPage
-: `RecordPage`s store the actual data. Currently, SirixDB stores 512 records in a record page. Each record page has a pointer to the previous record page, which stores the offset of the previous version of this record page. This pointer is crucial for our versioning algorithms, which have to retrieve several record pages (or record page fragments) to reconstruct a record page in-memory. Furthermore, records, which exceed a predefined size, are stored in so-called `OverflowPage`s and referenced in a record page.
+: `RecordPage`s store the actual data. Currently, SirixDB stores 512 records on a record page. Each record page has a pointer to the previous record page, which stores the offset of the previous version of this record page. This pointer is crucial for our versioning algorithms, which have to retrieve several record pages (or record page fragments) to reconstruct a record page in memory. Furthermore, records, which exceed a predefined size, are stored in so-called `OverflowPage`s and referenced on a record page.
```
Comment (mosheduminer) [inline concepts.md:77]: The correct usage here is `in`, not `on`, please change this back.

File: concepts.md [hunk 6]
```diff
@@ -7,118 +7,118 @@ title: SirixDB - Architecture and Concepts
...
 
 OverflowPage
 : `OverflowPage`s are used to store records, which exceeds a predefined size in bytes. When a transaction reads a record, SirixDB reconstructs the record page in-memory.
 
-However, potentially only a small fraction of records in the page have to be retrieved and reconstructed from byte-arrays in-memory for reads. SirixDB can delay this reconstruction until a transaction reads the overlong record.
+However, potentially only a small fraction of records on the page have to be retrieved and reconstructed from byte-arrays in-memory for reads. SirixDB can delay this reconstruction until a transaction reads the overlong record.
```
Comment (mosheduminer) [inline concepts.md:82]: Again, please change this back.

File: concepts.md (hunk 5 above)
Comment (mosheduminer) [reply concepts.md:77]: When dealing with a real page, you would be probably be correct, that it should be `on`, however, the conventional usage in regards to memory pages is `in`, I believe.

File: concepts.md [hunk 7]
```diff
@@ -7,118 +7,118 @@ title: SirixDB - Architecture and Concepts
...
 
-The next figure depicts what happens during a transaction-commit.
+The next figure depicts what happens during a transaction commit.
 
 <div class="img_container">
 ![pageStructure](images/copy-on-write.png){: style="max-width: 100%; height: auto; margin: 0em"}
 </div>
 
-We assume that a read-write transaction modifies a record in the leftmost *RecordPage*. Depending on the versioning algorithm used by SirixDB, the modified record, as well as probably some other records of the page, are copied to a new page fragment. First, SirixDB stores all changes in an in-memory transaction (intent) log. Second, during a transaction commit, the page-structure of the current *RevisionRootPage* is serialized in a postorder traversal.
+We assume that a read-write transaction modifies a record in the leftmost *RecordPage*. Depending on the versioning algorithm used by SirixDB, the modified record, as well as probably some other records of the page, are copied to a new page fragment. First, SirixDB stores all changes in an in-memory transaction (intent) log. Second, during a transaction commit, the page structure of the current *RevisionRootPage* is serialized in a postorder traversal.
```
Comment (mosheduminer) [inline concepts.md:92]: > page structure

given that structure here describes the page ("the page, which is a structure"), it's probably better to use `page-structure` .

File: concepts.md [hunk 8]
```diff
@@ -7,118 +7,118 @@ title: SirixDB - Architecture and Concepts
...
+All changed RecordPages are written to persistent storage, starting with the leftmost. If other changed record pages exist underneath an indirect page, SirixDB serializes these before the *IndirectPage*, which points to the updated record pages. Then the *IndirectPage* which points to the updated revision root page is written. The indirect pages are written with updated references to the new persistent locations of the record pages.
 
-SirixDB also stores checksums in the parent pointers as in ZFS. Thus, the storage engine in the future will be able to detect data corruption and heal itself, once we partition and replicate the data. SirixDB serializes the whole page-structure in this manner. We also intend to store an encryption key in the references in the future, to support encryption at rest.
+SirixDB also stores checksums in the parent pointers as in ZFS. Thus, the storage engine in the future will be able to detect data corruption and heal itself, once we partition and replicate the data. SirixDB serializes the whole page structure in this manner. We also intend to store an encryption key in the references in the future, to support encryption at rest.
```
Comment (mosheduminer) [inline concepts.md:96]: Again, please change this back.

File: concepts.md [hunk 9]
```diff
@@ -7,118 +7,118 @@ title: SirixDB - Architecture and Concepts
...
 
-All changed *RecordPages* are written to persistent storage, starting with the leftmost. If other changed record pages exist underneath an indirect page, SirixDB serializes these before the *IndirectPage*, which points to the updated record pages. Then the *IndirectPage* which points to the updated revision root page is written. The indirect pages are written with updated references to the new persistent locations of the record pages.
+All changed RecordPages are written to persistent storage, starting with the leftmost. If other changed record pages exist underneath an indirect page, SirixDB serializes these before the *IndirectPage*, which points to the updated record pages. Then the *IndirectPage* which points to the updated revision root page is written. The indirect pages are written with updated references to the new persistent locations of the record pages.
```
Comment (mosheduminer) [inline concepts.md:94]: Why remove the italics from `*RecordPages*`? Please put them back.

File: concepts.md [hunk 10]
```diff
@@ -7,118 +7,118 @@ title: SirixDB - Architecture and Concepts
...
+Note that SirixDB has to update the ancestor path of each changed RecordPage. However, storing indirect pages as well as the *RevisionRootPage*, *CASPage*, *PathSummaryPage*, and the *PathPage* is cheap. We currently store copies of the *NamePages*, but in the future might also version these according to the chosen versioning algorithm. Thus, we do not need to copy the whole dictionaries and save storage costs thereof. Each reference, which doesn’t point to a new page or page fragment, is left unchanged. Thus, unchanged pages (which are also not on the ancestor path of changed pages) are referenced at their respective position in the previous revision and never rewritten.
```
Comment (mosheduminer) [inline concepts.md:98]: Please put back the italics (`*`) and change back the apostrophe.

File: concepts.md [hunk 11]
```diff
@@ -7,118 +7,118 @@ title: SirixDB - Architecture and Concepts
...
 
 ### Versioning at the page-level
 
-One of the most distinctive features of SirixDB is that it versions the *RecordPages*. It doesn't merely copy all records of the page, even if a transaction only modifies a single record. The new record page fragment always contains a reference to the previous version. Thus, the versioning algorithms can dereference a fixed predefined number of page-fragments at maximum to reconstruct a RecordPage in-memory.
+One of the most distinctive features of SirixDB is that it versions the RecordPages. It doesn’t merely copy all records of the page, even if a transaction only modifies a single record. The new record page fragment always contains a reference to the previous version. Thus, the versioning algorithms can dereference a fixed predefined number of page-fragments at maximum to reconstruct a RecordPage in-memory.
```
Comment (mosheduminer) [inline concepts.md:102]: Please put back the italics (*) and change back the apostrophe.

File: concepts.md [hunk 12]
```diff
@@ -7,118 +7,118 @@ title: SirixDB - Architecture and Concepts
...
 
 **A sliding snapshot algorithm used to version record pages can avoid read and write peaks. The algorithm avoids intermittent full-page snapshots, which are otherwise needed during incremental or differential page-versioning to fast-track its reconstruction.**
 
 ### Versioning algorithms for storing and retrieving page snapshots
 
-SirixDB stores at most a fixed number of records. That is the actual data per database-page (currently limited to 512 records). The records themselves are of variable size. Overlong records, which exceed a predefined length in bytes, are stored in additional overflow pages. SirixDB stores references to these pages in the record pages.
+SirixDB stores at most a fixed number of records. That is the actual data per database page (currently limited to 512 records). The records themselves are of variable size. Overlong records, which exceed a predefined length in bytes, are stored in additional overflow pages. SirixDB stores references to these pages in the record pages.
```
Comment (mosheduminer) [inline concepts.md:108]: I think database-page is a particular kind of page, and warrants the use of the hyphen (`-`).

File: concepts.md [hunk 13]
```diff
@@ -7,118 +7,118 @@ title: SirixDB - Architecture and Concepts
...
 
-SirixDB implements several versioning strategies best known from backup systems for copy-on-write operations of record-pages. Namely it either copies
+SirixDB implements several versioning strategies best known from backup systems for copy-on-write operations of record-pages. Namely, it either copies
 
 - the full record-page that is any record in the page (full)
-- only the changed records in a record-page regarding the former version (incremental)
+- only the changed records in a record page regarding the former version (incremental)
```
Comment (mosheduminer) [inline concepts.md:113]: Please put the hyphen back (similar to database-page).

File: concepts.md [hunk 14]
```diff
@@ -7,118 +7,118 @@ title: SirixDB - Architecture and Concepts
...
 
 Incremental-versioning is one extreme. Write performance is best, as it stores the optimum (only changed records). On the other hand, reconstructing a page needs intermittent full snapshots of pages. Otherwise, performance deteriorates with each new revision of the page as the number of increments increases with each new version.
 
-Differential-versioning tries to balance reads and writes a bit better, but is still not optimal. A system, implementing a differential versioning strategy has to write all changed records since a past full dump of the page. Thus, only ever two revisions of the page-fragment have to be read to reconstruct a record-page. However, write-performance also deteriorates with each new revision of the page.
+Differential-versioning tries to balance reads and writes a bit better, but is still not optimal. A system, implementing a differential versioning strategy has to write all changed records since a past full dump of the page. Thus, only ever two revisions of the page fragment have to be read to reconstruct a record page. However, write performance also deteriorates with each new revision of the page.
```
Comment (mosheduminer) [inline concepts.md:118]: Please change record-page back.

File: concepts.md [hunk 15]
```diff
@@ -7,118 +7,118 @@ title: SirixDB - Architecture and Concepts
...
+The algorithm makes use of a sliding window. First, any changed record must be written during a commit. Second, any record, which is older than a predefined length N of the window and which has not been changed during these N-revisions must be written, too. Only these N-revisions at max have to be read. Fetching of the page fragments can be done in parallel or linear. In the latter case, the page fragments are read starting with the most recent revision. The algorithm stops once the full page has been reconstructed. You can find the best high-level overview of the algorithm in Marc’s Thesis: [Evolutionary Tree-Structured Storage: Concepts, Interfaces, and Applications](http://kops.uni-konstanz.de/handle/123456789/27695)
```
Comment (mosheduminer) [inline concepts.md:124]: Please change back the apostrophe.

**Review by mosheduminer — CHANGES_REQUESTED** [review]

File: concepts.md [hunk 16]
```diff
@@ -7,118 +7,118 @@ title: SirixDB - Architecture and Concepts
...
+SirixDB is a temporal database system and never overwrites data. Every time you’re committing a transaction, SirixDB creates a new lightweight snapshot. It uses a log-structured copy-on-write approach, whereas versioning takes place at the page- as well as node-level. Let’s first define what a temporal database system is all about.
 
-A temporal database is capable of retrieving past states. Typically it stores the transaction time; that is the time a transaction commits data. If the valid time is also stored, that is when a fact is true in the real world, we have a bitemporal relation, which is two time axes.
+A temporal database is capable of retrieving past states. Typically it stores the transaction time; that is the time a transaction commits data. If the valid time is also stored, that is when a fact is true in the real world, we have a bitemporal relation, which is two-time axes.
```
[the remark this answers is not in the document]
Comment (JohannesLichtenberger) [reply concepts.md:12]: Or maybe "which means two time axis", but I'm not sure about this.

**Comment by adnanmorsy97** [comment]
Hey @mosheduminer thank you for your hints. I edited everything, that you have mentioned. Could you please read it again and if everything is correct, could you merge it ?

**Pull request closed without merging**